### ISSUE TITLE
fix(Changedetection): ngAfterViewChecked is called when component is set to OnPush and no input properties changed

### ISSUE THREAD

**Issue by PascalPrecht** [angular/angular#7055]
Similar to https://github.com/angular/angular/issues/7054 it seems that `ngAfterViewChecked()` is called, when a component is set to `OnPush` and **no input property changed, or event is fired inside that component**.

Here's a plunk that demonstrates the issue: http://plnkr.co/edit/3HEH7AcrQaVUF0PTQBNv?p=preview

You'll see two components. One gets bootstrapped and the second one is set to `OnPush`, so it should only run change detection either one of its input properties changes, or an event is fired in side that component.

**Expectation**: OnPush Child Component should **not** turn green when the button is clicked, as no input properties are changing and no event is fired from inside the child component.

**Comment by choeller** [comment]
I also played around with that quite a bit, and I assume it's not _only_ the Callbacks getting fired, but that also some checks are made, that should not be. So in this plnkr: https://plnkr.co/edit/5mBfYdlluCWZammshsMR?p=preview

I have 3 Components - App, ContactList and ContactEntry. Like this:
```
@Component({
  selector: 'contact-entry',
  ...
 changeDetection: ChangeDetectionStrategy.OnPush
})
export class ContactEntry {
 ...
  checkCount = 0;
  ngDoCheck() {
    console.log(`Contact ${this.contact.name} checked ${this.checkCount++} times`);
  }
}
```
```
@Component({
  selector: 'contact-list',
  template: `
     <contact-entry *ngFor="#contact of contacts"
                   [contact]="contact"
                   (onSelect)="contactSelected(contact)">
     </contact-entry>
  `,
  directives: [ContactEntry],
  //changeDetection: ChangeDetectionStrategy.OnPush

})
export class ContactList {
checkCount = 0;
 ngDoCheck() {
    console.log(`ContactList checked ${this.checkCount++} times`);
  }
}
```
```
@Component({
  selector: 'app',
  template: `
      <h2>My Contacts</h2>
    <contact-list [contacts]="contacts"
                   (onSelect)="contactSelected($event)">
    </contact-list>
    <h3>Edit Contact</h3>
    <form>
      <label for="name">Name</label>
      <input type="text" id="name" [(ngModel)]="contactName">
      <label for="address">Address</label>
      <input type="text" id="address" [(ngModel)]="contactAddress">
      <button (click)="editContact()">Save</button>
      <button (click)="createContact()">Create new</button>
    </form>
  `,
  directives: [ContactList],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class App {
...
}
```

So basically a typical list with entries thing... When typing into the form, because of ngModel the Model of the app changes. 

Now when activating OnPush for the **List** the doCheck Callbacks of the List are called, but those of the **Entries** are not:
```
@Component({
  selector: 'contact-list',
  directives: [ContactEntry],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class ContactList {
checkCount = 0;
 ngDoCheck() {
    console.log(`ContactList checked ${this.checkCount++} times`);
  }
}
```
leads to the following logs when typing into the form of the app:
```
ContactList checked 1 times
app.ts:68 ContactList checked 2 times
app.ts:68 ContactList checked 3 times
app.ts:68 ContactList checked 4 times
app.ts:68 ContactList checked 5 times
app.ts:68 ContactList checked 6 times
app.ts:68 ContactList checked 7 times
app.ts:68 ContactList checked 8 times
app.ts:68 ContactList checked 9 times
...
```
What is interesting, is what happens, when I remove the OnPush from the list - then also the check-callbacks of the ContactEntry Instances are called:
```
@Component({
  selector: 'contact-list',
  directives: [ContactEntry],
//  changeDetection: ChangeDetectionStrategy.OnPush
})
export class ContactList {
```
```
ContactList checked 1 times
app.ts:41 Contact John Doe checked 1 times
app.ts:41 Contact Richard Roe checked 1 times
app.ts:68 ContactList checked 2 times
app.ts:41 Contact John Doe checked 2 times
app.ts:41 Contact Richard Roe checked 2 times
app.ts:68 ContactList checked 3 times
app.ts:41 Contact John Doe checked 3 times
app.ts:41 Contact Richard Roe checked 3 times
app.ts:68 ContactList checked 4 times
app.ts:41 Contact John Doe checked 4 times
app.ts:41 Contact Richard Roe checked 4 times
...
```
So neither a binding of the list nor a binding of the Entries are changed, but switching Strategy for the List influences the Checks of the Entries, so IMHO this shows, that the problem is not only about triggering the callback. Or am I having an error in reasoning here? Also each keystroke triggers ngDoCheck-callback for each component twice, but I guess this is another problem...

**Comment by DzmitryShylovich** [comment]
4.0 http://plnkr.co/edit/lU32vBJY9KYXZUTW7GQQ?p=preview

**Comment by illiano** [comment]
I have an example of two sibling components, where even when nothing changes inside the components the 'ngAfterViewChecked' will called

http://plnkr.co/edit/eFtn0PZhEm6ppoZGvvr8?p=preview

**Comment by jredd** [comment]
I have an example as well when using a url sanitizer. It keeps reloading the pdf that i'm adding to the page anytime an angular action takes place (mouseout, mousenter), even though the url doesn't change at all.

https://plnkr.co/edit/sovRUOn79LTJRDhaellf?p=preview

**Comment by acheshkov** [comment]
v2.4.0. the problem exists

**Comment by MoiraG** [comment]
I had tried to understand Angular's Changedetection and was surprised that ngAfterViewChecked is called, although I had changed the ChangeDetectionStrategy to onPush. It seems to me that the component is not really being checked. Here is a small example with 2 components: https://stackblitz.com/edit/angular-id7pju

**Comment by mlc-mlapis** [comment]
@MoiraG ... there is a sequence of hooks `ngOnChanges() -> ngDoCheck() -> ngAfterContentChecked() -> ngAfterViewChecked()` which are processed in any case -> CD. That is why it could be so expensive to place any code into them ... because it runs so frequently. The hook `ngOnChanges` also decides if a component should be refreshed or not ... and it is also true for `OnPush`.

**Comment by MoiraG** [comment]
in the example I tried to show that `ngAfterViewChecked `is called when a component is set to OnPush and no input property has been changed and no event has been triggered within this component.

Is that expected behavior?

**Comment by mlc-mlapis** [comment]
@MoiraG ... yes, this is expected ... your example also shows that the component refresh is not done ... just hooks `ngOnChanges()` -> `ngDoCheck()` -> `ngAfterContentChecked()` -> `ngAfterViewChecked()` are processed ... and it is fine because there is zero performance hit. The opposite situation would be if you pass a new reference through the `@Input()` (as OnPush expects) ... then `ngOnChanges()` detects that a full refresh of the component should be done ... and this IS a potential performance hit.

**Comment by MoiraG** [comment]
@mlc-mlapis Ok thanks for the information. I thought ngAfterViewChecked is only called if the component is really checked.

**Comment by binDongKim** [comment]
@mlc-mlapis This doesn't sound make sense to me because based on my understanding, the component with the `OnPush` strategy is supposed to be checked when its input property changes so if any change doesn't happen, the check shouldn't be done on the component, which leads me to think that `ngAfterViewChecked` shouldn't be called as well. What am I missing?

**Comment by mlc-mlapis** [comment]
@binDongKim ... hmm, `ngDoCheck()` on a child component is invoked from its parent component side (not from a child component itself) on each CD and it doesn't matter if the child has or doesn't `OnPush` strategy enabled. **This is the only way how to detect some mutable changes on an object passed as `@Input()` to a child when `OnPush` is enabled** ... when you broke the `immutability` principle from any reason.

Generally it is recommended to be very careful when adding anything to this hook and it is the same for `ngAfterContentChecked()` which is called after `ngDoCheck()` and `ngAfterViewChecked()` which is called after `ngAfterContentChecked()`. All these 3 hooks are the only places where you can force to run your code even in the cases when Angular is applying its own principles and controls its flow.

**Comment by binDongKim** [comment]
@mlc-mlapis Got it! Thanks for your explanation. I really appreciate it.

**Comment by ChristophWieske** [comment]
@mlc-mlapis I don´t see why it should be okay for ngAfterViewChecked to be called when the view wasn´t checked.

If I (for what ever reason I want to do this - maybe a demo on how the change detection works) hook up on ngAfterViewChecked to do something after the change detection ran - and only then - this would fail. Indicating that there were CD where it shouldn´t have.

According to the docs ngAfterViewChecked should be called after CD checked the components view. So if it wasn´t checked...

@PascalPrecht saw your talk at NGNL2016: Can´t believe this is still an issue.

**Comment by mlc-mlapis** [comment]
@ChristophWieske ... I suggest reading the following post: https://blog.angularindepth.com/a-gentle-introduction-into-change-detection-in-angular-33f9ffff6f10 from our dear Max. Especially the section:

> Let’s go through it step by step. First, it updates the input bindings for the child component. Then it calls the OnInit, DoCheck, and OnChanges hooks, again, on the child component. It makes sense because it just updated the input bindings and Angular needs to notify the child components that the input bindings have been initialized. Then Angular performs rendering for the current component. And after that, it runs change detection for the child component. This means that it’ll basically repeat these operations on the child view. And finally, it calls theAfterViewChecked and AfterViewInit hooks on the child component to let it know that it’s been checked.

![image](https://user-images.githubusercontent.com/5693835/55837874-e88bc300-5b22-11e9-98d8-2ed5cce4a2b6.png)

**Comment by ChristophWieske** [comment]
Thank you for your quick response @mlc-mlapis . That article is indeed pretty good, but the mistake is still the same.

> And finally, it calls theAfterViewChecked and AfterViewInit hooks on the child component to let it know that it’s been checked
--> it hasn´t been checked! Even though the child components data have been changed, its not projected. There is no check.

If it would have been checked, everything would be fine. But this is not the case. There is no CD executed...

**Comment by mlc-mlapis** [comment]
@ChristophWieske ... I think that you are fixed to the actual naming too much. It's not perfect for sure.

Probably the correct definition should be: **It calls the AfterViewChecked hook on the child component to let it know that it’s been checked. Even if there was no necessity to run the CD cycle on it, the calling of AfterViewChecked still means the fact that it happend in a controlled way and it was checked that.**

**Comment by ChristophWieske** [comment]
@mlc-mlapis I was very close to reply that there was no check at all for the child components. If so a change to the data of the child component I made from my parent components click handler should have been noticed and have forced the CD to run... yeah but I have `OnPush` set... and I didn´t used a new object reference.

I updated my sample and... yes, you were right!

The check whether something changed happens the way it should with OnPush. So okay, its a docs issue not really a bug. As the view was checked (to decide whether the DOM needs to be checked... many checks in here!) the naming is probably okay.

Thanks for clarifying.

Besides: Is there a way on getting noticed when the change detection **really** runs? I prepare a talk and try to make a demo for this. I would like to differenciate those actions. Do you have any idea on how to achieve this?

**Comment by mlc-mlapis** [comment]
@ChristophWieske ... do you mean that still in the context of `OnPush` strategy?

**Comment by ChristophWieske** [comment]
@mlc-mlapis yes... as checking the view and running CD are two different things (or better stages) I want to show that to the audience.

**Comment by mlc-mlapis** [comment]
@ChristophWieske ... so because of `OnPush` strategy and its functionality depend strictly on `@Input()` changes, I use the code as bellow to detect and check during development if and what invoked CD on components with `OnPush` strategy:

```
ngOnChanges(changes: SimpleChanges): void {
	if (changes['rowClassInfo']) {
		if (!changes['anyInputPropertyName'].isFirstChange()) {
			console.log('... changes on the ... [anyInputPropertyName]:', ...);
		}
	}
}
```

**Comment by ChristophWieske** [comment]
For when the CD is input triggered that would be cool. But In my case a click event will trigger it marking the component and its ancestors dirty without an input changed. Will that fire for the dirty-marked ancestors as well?

I found a different (and shockingly simple) approach: using a getter bound to a noop directive in the view. Whenever the getter is invoked the CD is running for it. Found it [here](https://www.lucidchart.com/techblog/2018/07/02/seeing-angular-change-detection-in-action-part-ii-onpush-observables-and-the-async-pipe/).

Could be worth a look!

@mlc-mlapis thank you for being my sparing partner to this issue. You get a special thanks on my slides.

**Comment by mlc-mlapis** [comment]
@ChristophWieske ... nice to hear, thank you! 😄 Btw ... the article is a really good one.

**Comment by bencody** [comment]
I'm also confused by Angular's change detection system and its relation to the component lifecycle methods. The official documentation isn't clear on how OnPush is different from Default, and the description of the lifecycle methods also isn't great.

I expected that making my components OnPush-based would prevent their ngAfterViewChecked() methods from being called repeatedly when their inputs don't change, but as @ChristophWieske also noticed, that isn't the case. I had hoped that I could use ngAfterViewChecked() as a simple proof that switching my components to OnPush reduces the amount of processing that is done on them.

Are there any good profiling tools for Angular that can clearly show you the difference in performance between OnPush and the default change detection system?

**Comment by mlc-mlapis** [comment]
@bencody The `OnPush` strategy is a straightforward thing. It merely detects if any object reference passed as input has been changed or not + any change of primitive input values. That's all. If that happened, the CD is processed on the component instance with all effects on the UI template. If not, CD is not processed.

**Comment by muuvmuuv** [comment]
I want to add my use case because I run out of ideas: I have a component called "card" which is dynamic and has a directive in it. I can pass a type prop which will create a component via `ComponentFactoryResolver` inside the directive to create the correct "card" component. Inside the card I have data that needs to be resolved. For example the card can have a prop location which is different based on the type, so I have a function that returns the correct translation string based on those two: `{{ getLocationText() }}`. The issue here is that because of the multiple times called `ngAfterContentChecked` and `ngAfterViewChecked` the function is called 30 times. I have no idea if this adds performance costs but I guess it does. I want to prevent that since nothing has changed and the card is just initialized.

Here is a peek-a-boo component to visualize it a little:

```
ComponentFactory$1 {componentDef: {…}, ngModule: NgModuleRef$1, selector: "app-card-activity-default", ngContentSelectors: Array(0), componentType: ƒ, …}
peek-a-boo.component.ts:29 CONSTRUCTED: CardActivity
peek-a-boo.component.ts:42 [CardActivity] => Before rendering: ngOnInit
peek-a-boo.component.ts:45 [CardActivity] => Before rendering: ngDoCheck
peek-a-boo.component.ts:56 [CardActivity] => Change detection: ngAfterContentInit
peek-a-boo.component.ts:68 [CardActivity] => After rendering: ngAfterContentChecked
default.component.ts:23 FUNC_TRIGGERED
peek-a-boo.component.ts:59 [CardActivity] => Change detection: ngAfterViewInit
peek-a-boo.component.ts:65 [CardActivity] => After rendering: ngAfterViewChecked
peek-a-boo.component.ts:45 [CardActivity] => Before rendering: ngDoCheck
peek-a-boo.component.ts:68 [CardActivity] => After rendering: ngAfterContentChecked
default.component.ts:23 FUNC_TRIGGERED
peek-a-boo.component.ts:65 [CardActivity] => After rendering: ngAfterViewChecked
default.component.ts:23 FUNC_TRIGGERED
peek-a-boo.component.ts:45 [CardActivity] => Before rendering: ngDoCheck
peek-a-boo.component.ts:68 [CardActivity] => After rendering: ngAfterContentChecked
default.component.ts:23 FUNC_TRIGGERED
peek-a-boo.component.ts:65 [CardActivity] => After rendering: ngAfterViewChecked
default.component.ts:23 FUNC_TRIGGERED
peek-a-boo.component.ts:45 [CardActivity] => Before rendering: ngDoCheck
peek-a-boo.component.ts:68 [CardActivity] => After rendering: ngAfterContentChecked
default.component.ts:23 FUNC_TRIGGERED
peek-a-boo.component.ts:65 [CardActivity] => After rendering: ngAfterViewChecked
default.component.ts:23 FUNC_TRIGGERED
peek-a-boo.component.ts:45 [CardActivity] => Before rendering: ngDoCheck
peek-a-boo.component.ts:68 [CardActivity] => After rendering: ngAfterContentChecked
default.component.ts:23 FUNC_TRIGGERED
peek-a-boo.component.ts:65 [CardActivity] => After rendering: ngAfterViewChecked
default.component.ts:23 FUNC_TRIGGERED
peek-a-boo.component.ts:45 [CardActivity] => Before rendering: ngDoCheck
peek-a-boo.component.ts:68 [CardActivity] => After rendering: ngAfterContentChecked
default.component.ts:23 FUNC_TRIGGERED
peek-a-boo.component.ts:65 [CardActivity] => After rendering: ngAfterViewChecked
default.component.ts:23 FUNC_TRIGGERED
peek-a-boo.component.ts:45 [CardActivity] => Before rendering: ngDoCheck
peek-a-boo.component.ts:68 [CardActivity] => After rendering: ngAfterContentChecked
default.component.ts:23 FUNC_TRIGGERED
peek-a-boo.component.ts:65 [CardActivity] => After rendering: ngAfterViewChecked
default.component.ts:23 FUNC_TRIGGERED
peek-a-boo.component.ts:45 [CardActivity] => Before rendering: ngDoCheck
peek-a-boo.component.ts:68 [CardActivity] => After rendering: ngAfterContentChecked
default.component.ts:23 FUNC_TRIGGERED
peek-a-boo.component.ts:65 [CardActivity] => After rendering: ngAfterViewChecked
default.component.ts:23 FUNC_TRIGGERED
peek-a-boo.component.ts:45 [CardActivity] => Before rendering: ngDoCheck
peek-a-boo.component.ts:68 [CardActivity] => After rendering: ngAfterContentChecked
default.component.ts:23 FUNC_TRIGGERED
peek-a-boo.component.ts:65 [CardActivity] => After rendering: ngAfterViewChecked
default.component.ts:23 FUNC_TRIGGERED
peek-a-boo.component.ts:45 [CardActivity] => Before rendering: ngDoCheck
peek-a-boo.component.ts:68 [CardActivity] => After rendering: ngAfterContentChecked
default.component.ts:23 FUNC_TRIGGERED
peek-a-boo.component.ts:65 [CardActivity] => After rendering: ngAfterViewChecked
default.component.ts:23 FUNC_TRIGGERED
peek-a-boo.component.ts:45 [CardActivity] => Before rendering: ngDoCheck
peek-a-boo.component.ts:68 [CardActivity] => After rendering: ngAfterContentChecked
default.component.ts:23 FUNC_TRIGGERED
peek-a-boo.component.ts:65 [CardActivity] => After rendering: ngAfterViewChecked
default.component.ts:23 FUNC_TRIGGERED
peek-a-boo.component.ts:45 [CardActivity] => Before rendering: ngDoCheck
peek-a-boo.component.ts:68 [CardActivity] => After rendering: ngAfterContentChecked
default.component.ts:23 FUNC_TRIGGERED
peek-a-boo.component.ts:65 [CardActivity] => After rendering: ngAfterViewChecked
default.component.ts:23 FUNC_TRIGGERED
peek-a-boo.component.ts:45 [CardActivity] => Before rendering: ngDoCheck
peek-a-boo.component.ts:68 [CardActivity] => After rendering: ngAfterContentChecked
default.component.ts:23 FUNC_TRIGGERED
peek-a-boo.component.ts:65 [CardActivity] => After rendering: ngAfterViewChecked
default.component.ts:23 FUNC_TRIGGERED
client:48 [WDS] Hot Module Replacement enabled.
client:52 [WDS] Live Reloading enabled.
```

**Comment by mlc-mlapis** [comment]
@muuvmuuv Do you have any functional code inside those two hooks?

**Comment by muuvmuuv** [comment]
@mlc-mlapis no, I just added it so I can see when and where after my function is called. Anyway it is called too often.

**Comment by mlc-mlapis** [comment]
@muuvmuuv OK. Not in reality, because of Angular compiler creates the magic when compiling the code, and if it doesn't find that hook, it doesn't create any code that can be called, and this means the zero load. Btw, `ngDoCheck()`, `ngAfterContentChecked()`, `ngAfterViewChecked()` are a bit special hooks, because they allow you to overwrite the default behavior, even for `OnPush` strategy, so if they need do that, they have to be always run.

**Comment by muuvmuuv** [comment]
@mlc-mlapis I understand that, but anyway when I remove all hooks, It is still the same. Function is called 20 times. Maybe I should create another issue for that? I am pretty sure it is the factory resolver or container ref which vreates the dynamic component.

**Comment by mlc-mlapis** [comment]
@muuvmuuv How can be called when it doesn't exist?

**Comment by muuvmuuv** [comment]
@mlc-mlapis it does? I don‘t understand. I said when I remove all hooks. The function in default.component.ts is called in the template not in a hook. I just added them to better understand what is going on.

**Comment by mlc-mlapis** [comment]
@muuvmuuv Ah, you mean that `{{ getLocationText() }}` is called multiple times, right? You have to differentiate between JIT and AOT mode of the app then. In JIT, it is called two times more than in AOT because the JIT logic does some additional checks. But still, it is called twice for each instance of such a binding. That's why binding to functions in a template is generally not recommended, and if used, it's necessary to understand what you are doing and how that function code is performant. Calling a function as binding is a part of CD logic in the whole hierarchy tree of components. There are other more performant code constructs for such cases.

**Comment by muuvmuuv** [comment]
Thank you for the detailed explanation. This makes more sense now. We use AOT (because it was enabled by default in ng9). I have now moved this function into `ngOnInit` and declare a variable which replaces the function `{{ locationText }}`. Feels a little faster now.

**Comment by mlc-mlapis** [comment]
@muuvmuuv Sure, because of only a static property reference in bonded. You can think also about using pure pipes in such situations.

**Comment by TheParad0X** [comment]
So... is this a bug or not? The hook is called even if the CD did not happen for my OnPush component? 🤔

**Comment by mlc-mlapis** [comment]
@TheParad0X The `ngAfterViewChecked` is not related to OnPush. It's a way how to allow to run custom code in any circumstances and because of that, you should be very strict what code and why you place it there.

**Comment by TheParad0X** [comment]
@mlc-mlapis I find it confusing that ngAfterViewChecked is called even if the CD did NOT check my view (whose attached component is OnPush). I cannot use this hook to show if the CD has run or not.

**Comment by ahmedayman94** [comment]
My understanding of this, is that the strategy affects what is checked.
OnDefault, all model bindings are checked for a given component, no questions asked.

OnPush, only input bindings to this component are checked. If input binding have changed, then all model bindings for that given component will be checked, and subsequent update of DOM will follow. However, if it's an event that is triggered by this component (click on button for e.g), then CD will execute on all bindings.

In the Parent-Child relationship, a parent will trigger CD on its child, who based on this child's CD strategy, will determine what is to be checked. In both cases, CD runs but the difference between what determines what gets checked is the strategy.
Default will check all bindings of the child, in OnPush it will only check the input bindings.

Is this correct ?

**Comment by mlc-mlapis** [comment]
@ahmedayman94 Simple demo: https://stackblitz.com/edit/angular-parent-child-onpush?file=src%2Fapp%2Fparent%2Fparent.component.ts

**Comment by ahmedayman94** [comment]
thanks for the reply Milos.

I think this supports my understanding that CD either runs on all model bindings of a given component or just the input for it self (in case of onPush) to determine whether it needs to check all bindings, the initiator of that action would be the parent.

Just checking the inputs to child-component triggers the ngAfterViewChecked for this child even tho it technically didnt check all bindings it has in the child-template, and therefore it might go out of sync.

That was the part that had confused me, afterViewChecked lifecycle hook triggering for a component whose bindings have not been updated.

So the best way to know if change detection has triggered for the bindings of a given component is not the ngAfterViewChecked, but instead having a getter and using its value in its template, a console log in that would fire if a proper CD on its bindings have triggered, and i think that's the one in question when it comes to performance benefits, because the more bindings you have, the more expensive this operation would be and that's what you would try to minimize by having the check only on the input to determine if all bindings need a check in the first place.

The terminology is a bit confusing for me. Let me know if I misunderstood something.

**Comment by mlc-mlapis** [comment]
@ahmedayman94 Yep, `ngAfterViewChecked` is not the best place to confirm the CD. One extra fact: the `ngDoCheck` hook always runs, so you can use it to overwrite the standard `OnPush` behavior if you want to do it under some extra wanted logic.

**Comment by david-knipe395** [comment]
> @ChristophWieske ... I suggest reading the following post: https://blog.angularindepth.com/a-gentle-introduction-into-change-detection-in-angular-33f9ffff6f10 from our dear Max. Especially the section:

That link is broken. It's now here: https://angularindepth.com/posts/1058/a-gentle-introduction-into-change-detection-in-angular
